// @getmocha/vite-plugins/dist/mocha-watermark/mocha-watermark-injected.js
function shouldShowWatermark() {
    if (MOCHA_ENV.SHOW_WATERMARK !== 'true' && MOCHA_ENV.SHOW_WATERMARK !== true) {
        return false;
    }
    if (window.top !== window && isSameDomain(document.referrer, MOCHA_ENV.APP_URL ?? '')) {
        return false;
    }
    return true;
}
function isSameDomain(url1, url2) {
    try {
        const u1 = new URL(url1);
        const u2 = new URL(url2);
        return u1.origin === u2.origin;
    }
    catch {
        return false;
    }
}
if (shouldShowWatermark()) {
    const style = document.createElement('style');
    style.textContent = `
    .mocha-watermark {
      position: fixed;
      bottom: 16px;
      right: 16px;
      background: white;
      border-radius: 8px;
      box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
      display: flex;
      align-items: center;
      padding: 8px 12px;
      z-index: 9999;
      font-family: "Instrument Sans", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif;
      font-size: 14px;
      font-weight: 600;
      color: #000;
      gap: 8px;
      border: 1px solid #e6e6e6;
      background: linear-gradient(to bottom, #FFFFFF, #F9F9F9);
      cursor: pointer;
      transition: all 0.2s ease-in-out;
    }

    .mocha-watermark:hover {
      transform: translateY(-2px);
      box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    }

    .mocha-watermark:active {
      transform: translateY(0);
      box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
    }

    .mocha-watermark img {
      width: 20px;
      height: 20px;
    }
  `;
    const watermark = document.createElement('a');
    const baseUrl = MOCHA_ENV.APP_URL ?? 'https://getmocha.com';
    watermark.href = `${baseUrl}?a_id=${MOCHA_ENV.APP_ID}`;
    watermark.target = '_blank';
    watermark.className = 'mocha-watermark';
    watermark.innerHTML = `
    <img src="https://static.getmocha.com/favicon.svg" alt="Mocha Logo" />
    Clone with Mocha
  `;
    document.head.appendChild(style);
    document.body.appendChild(watermark);
}
export {};
