// @getmocha/vite-plugins/dist/error-reporter/error-reporter.js
import { injectedClientModule } from '../injectedClientModule.js';
export function errorReporter(env) {
    return {
        name: 'mocha-error-reporter',
        enforce: 'pre',
        // See: https://vite.dev/guide/using-plugins#conditional-application
        apply: 'serve',
        configureServer(server) {
            server.middlewares.use((req, res, next) => {
                const url = req.url || '';
                res.on('finish', () => {
                    const { statusCode, statusMessage } = res;
                    const isOutdatedOptimizeDepError = statusCode === 504 && statusMessage?.toLowerCase().includes('outdated optimize dep');
                    if (!isOutdatedOptimizeDepError)
                        return;
                    if (env.DEBUG_LOGS === 'true') {
                        console.log('[mocha-error-reporter] optimized dependencies changed. reloading');
                    }
                    server.ws.send({
                        type: 'custom',
                        event: 'vite:outdatedOptimizeDeps',
                        data: { url: url },
                    });
                    server.ws.send({ type: 'full-reload' });
                });
                next();
            });
        },
        ...injectedClientModule('error-reporter', env, { injectTo: 'head-prepend' }),
    };
}
